Service/AlgoliaConnector.php

```diff
@@ -0,0 +1,984 @@
+            }
+            return [];
+        }
+    }
+
+    /**
+     * @param $indexName
+     * @param $settings
+     * @param string $mergeSettingsFrom
+     * @return SettingsResponse|array
+     */
+    public function mergeSettings($indexName, $settings, string $mergeSettingsFrom = ''): SettingsResponse|array
+    {
+        $onlineSettings = [];
+
+        try {
+            $sourceIndex = $indexName;
+            if ($mergeSettingsFrom !== '') {
+                $sourceIndex = $mergeSettingsFrom;
+            }
+
+            $onlineSettings = $this->getClient()->getSettings($sourceIndex);
+        } catch (Exception $e) {
+        }
+
+        if (isset($settings['attributesToIndex'])) {
+            $settings['searchableAttributes'] = $settings['attributesToIndex'];
+            unset($settings['attributesToIndex']);
+        }
+
+        if (isset($onlineSettings['attributesToIndex'])) {
+            $onlineSettings['searchableAttributes'] = $onlineSettings['attributesToIndex'];
+            unset($onlineSettings['attributesToIndex']);
+        }
+
+        foreach ($this->getSettingsToRemove($onlineSettings) as $remove) {
+            if (isset($onlineSettings[$remove])) {
+                unset($onlineSettings[$remove]);
+            }
+        }
+
+        foreach ($settings as $key => $value) {
+            $onlineSettings[$key] = $value;
+        }
+
+        return $onlineSettings;
+    }
+
+    /**
+     * These settings are to be managed by other processes
+     * @param string[] $onlineSettings
+     * @return string[]
+     */
+    protected function getSettingsToRemove(array $onlineSettings): array
+    {
+        $removals = ['slaves', 'replicas', 'decompoundedAttributes'];
+
+        if (isset($onlineSettings['mode']) && $onlineSettings['mode'] == 'neuralSearch') {
+            $removals[] = 'mode';
+        }
+
+        return array_merge($removals, $this->getSynonymSettingNames());
+    }
+
+    /**
+     * @return string[]
+     */
+    protected function getSynonymSettingNames(): array
+    {
+        return [
+            'synonyms',
+            'altCorrections',
+            'placeholders'
+        ];
+    }
+
+    /**
+     * Legacy function signature to add objects to Algolia
+     * @param array $objects
+     * @param string $indexName
+     * @return void
+     * @throws Exception
+     * @deprecated Do not use. This method has been replaced by saveObjects and may be removed in the future.
+     */
+    public function addObjects(array $objects, string $indexName): void {
+        $this->saveObjects($indexName, $objects, $this->config->isPartialUpdateEnabled());
+    }
+
+    /**
+     * Save objects to index (upserts records)
+     * @param string $indexName
+     * @param array $objects
+     * @param bool $isPartialUpdate
+     * @return void
+     * @throws Exception
+     */
+    public function saveObjects(string $indexName, array $objects, bool $isPartialUpdate = false): void
```

Comment:
What do you think about this idea?... if we were to make this truly stateless and accept `$storeId` for these functions then we don't need `$indexName`. That can be obtained from something like the `IndexNameFetcher`. 

(Of course if we are working with temp indices then that needs to be extrapolated differently. For this we could pass a `$isTempIndex` boolean or even more flexibly an array of options or even some kind of `IndexOptions` object.)

The reason I care about this is because there is the potential for conflict between `$storeId` and `$indexName` and you could conceivably pass an index name that doesn't belong to the store ID. I know this is a misuse of the API but I like the idea of making this more foolproof. It was also the reason I took index names out of the `ReplicaManager` implementation.

This is also a bit more ambitious so if we can agree on the most practical approach maybe we break that up as a separate task.

Reply:
See comment on JIRA :)